titan doc update for 7.6 release (#15954)
pingcap/docs-cn: storage-engine/titan-configuration.md

# Titan 配置

本文档介绍如何通过 [Titan](/storage-engine/titan-overview.md) 配置项来开启、关闭 Titan、数据迁移原理、相关参数以及 Level Merge 功能。

## 开启 Titan

> **注意：**
>

> > - 从 TiDB v7.6.0 开始，新集群将默认启用 Titan，并将阈值 [`min-blob-size`](/tikv-configuration-file.md#min-blob-size) 的默认值由之前的 `1KB` 调整为 `32KB`。

Reviewer: ```suggestion
> - 从 TiDB v7.6.0 开始，新集群将默认启用 Titan，以更好地支持 TiDB 宽表写入场景，特别是在支持 JSON 之后。阈值 [`min-blob-size`](/tikv-configuration-file.md#min-blob-size) 的默认值也由之前的 `1KB` 调整为 `32KB`。
```

    ```shell
    tiup cluster reload ${cluster-name} -R tikv
    ```

    具体命令，可参考[通过 TiUP 修改配置参数](/maintain-tidb-using-tiup.md#修改配置参数)。

+ 方法二：直接编辑 TiKV 配置文件开启 Titan（不建议在生产环境中使用）。

    ```toml
    [rocksdb.titan]
    enabled = true
    ```

+ 方法三：编辑 TiDB Operator 的 `${cluster_name}/tidb-cluster.yaml` 配置文件，编辑示例如下：

    ```yaml
    spec:
      tikv:
        ## Base image of the component
        baseImage: pingcap/tikv
        ## tikv-server configuration
        ## Ref: https://docs.pingcap.com/zh/tidb/stable/tikv-configuration-file
        config: |
          log-level = "info"
          [rocksdb]
            [rocksdb.titan]
              enabled = true
    ```

    应用配置时，触发在线滚动重启 TiDB 集群让配置生效：

    ```shell
    kubectl apply -f ${cluster_name} -n ${namespace}
    ```

    更多信息请参考[在 Kubernetes 中配置 TiDB 集群](https://docs.pingcap.com/zh/tidb-in-kubernetes/stable/configure-a-tidb-cluster)。

## 数据迁移

> **警告：**
>
> 在关闭 Titan 功能的情况下，RocksDB 无法读取已经迁移到 Titan 的数据。如果在打开过 Titan 的 TiKV 实例上错误地关闭了 Titan（误设置 `rocksdb.titan.enabled = false`），启动 TiKV 会失败，TiKV log 中出现 `You have disabled titan when its data directory is not empty` 的错误。如需要关闭 Titan，请参考[关闭 Titan](#关闭-titan) 。

> 开启 Titan 以后，原有的数据并不会马上迁移到 Titan 引擎，而是随着前台写入和 RocksDB Compaction 的进行，**逐步进行 key-value 分离并写入 Titan**。同样的，无论是通过 [BR](/br/backup-and-restore-overview.md) 快照或日志恢复的数据，还是通过 [TiDB Lightning](/tidb-lightning/tidb-lightning-overview.md) 逻辑导入模式或物理导入模式导入的数据，都会首先写入 RocksDB。然后，随着 RocksDB Compaction 的进行，超过 [`min-blob-size`](/tikv-configuration-file.md#min-blob-size) 默认值 `32KB` 的大 value 会逐步分离到 Titan 中。你可以通过观察 **TiKV Details** - **Titan kv** - **blob file size** 监控面板中文件的大小来确认存储在 Titan 中的数据大小。

Reviewer 2: lightning 逻辑导入应该是 sql 写入的形式

为了更快地将数据转移到 Titan，建议使用 tikv-ctl 工具执行一次全量 Compaction，以提高迁移速度。具体操作步骤请参考[手动 compact](/tikv-control.md#手动-compact-整个-tikv-集群的数据)。由于 RocksDB 具备 Block Cache，并且在将数据从 RocksDB 迁移到 Titan 时，数据访问是连续的，这使得在迁移过程中 Block Cache 能够更有效地提升迁移速度。在我们的测试中，通过 tikv-ctl 在单个 TiKV 节点上执行全量 Compaction，仅需 1 小时就能将 670 GiB 的数据迁移到 Titan。

需要注意的是，由于 Titan Blob 文件中的 Value 并非连续的，而且 Titan 的缓存是基于 Value 级别的，因此 Blob Cache 无法在 Compaction 过程中提供帮助。相较于从 RocksDB 转向 Titan 的速度，从 Titan 转回 RocksDB 的速度则会慢一个数量级。在测试中，通过 tikv-ctl 将 TiKV 节点上的 800 GiB Titan 数据进行全量 Compaction 转为 RocksDB，需要花费 12 个小时。

## 常用配置参数

通过合理配置 Titan 参数，可以有效提升数据库性能和资源利用率。

### `min-blob-size`

> 你可以通过设置 [`min-blob-size`](/tikv-configuration-file.md#min-blob-size) 来调整 value 的大小阈值，决定哪些数据保存在 RocksDB 中，哪些数据保存在 Titan 的 blob file 中。经过[测试](/storage-engine/titan-overview.md#min-blob-size-的选择及其性能影响)，`32KB` 是个折中的值。

Reviewer: ```suggestion
你可以通过设置 [`min-blob-size`](/tikv-configuration-file.md#min-blob-size) 来调整 value 的大小阈值，决定哪些数据保存在 RocksDB 中，哪些数据保存在 Titan 的 blob file 中。经过[测试](/storage-engine/titan-overview.md#min-blob-size-的选择及其性能影响)，`32KB` 是个折中的值。如果你想进一步提升写性能，并能接受扫描性能的下降，你可以将该值调整为 `1KB`。
```

### `blob-file-compression` 和 `zstd-dict-size`

可以使用 [`blob-file-compression`](/tikv-configuration-file.md#blob-file-compression) 参数指定 Titan 中 value 所使用的压缩算法，也可以配置 [`zstd-dict-size`](/tikv-configuration-file.md#zstd-dict-size) 启用 `zstd` 字典压缩来提高压缩率。

### `blob-cache-size`

> 要注意的是，[`blob-cache-size`](/tikv-configuration-file.md#blob-cache-size) 控制 Titan 中 value 的缓存大小。更大的缓存能提高 Titan 读性能，但过大的缓存会造成 OOM。建议在数据库稳定运行后，根据监控把 RocksDB block cache (`storage.block-cache.capacity`) 设置为 store size 减去 blob file size 的大小，`blob-cache-size` 设置为内存大小 * 50% 减去 block cache 的大小。这是为了保证 block cache 足够缓存整个 RocksDB 的前提下，blob cache 尽量大。

Reviewer: ```suggestion
可以使用 [`blob-cache-size`](/tikv-configuration-file.md#blob-cache-size) 控制 Titan 中 value 的缓存大小。更大的缓存能提高 Titan 读性能，但过大的缓存会造成 OOM。建议在数据库稳定运行后，根据监控把 RocksDB block cache (`storage.block-cache.capacity`) 设置为 store size 减去 blob file size 的大小，`blob-cache-size` 设置为内存大小 * 50% 减去 block cache 的大小。这是为了保证 block cache 足够缓存整个 RocksDB 的前提下，blob cache 尽量大。
```

### `discardable-ratio` 和 `max-background-gc`

> [`discardable-ratio`](/tikv-configuration-file.md#discardable-ratio) 和 [`max-background-gc`](/tikv-configuration-file.md#max-background-gc) 的设置对于 Titan 的读性能和垃圾回收过程都有重要影响。当一个 blob file 中无用数据（相应的 key 已经被更新或删除）比例超过 [`discardable-ratio`](/tikv-configuration-file.md#discardable-ratio) 设置的阈值时，将会触发 Titan GC。减少这个阈值可以减少空间放大，但是会造成 Titan 更频繁 GC；增加这个值可以减少 Titan GC，减少相应的 I/O 带宽和 CPU 消耗，但是会增加磁盘空间占用。

Reviewer: ```suggestion
[`discardable-ratio`](/tikv-configuration-file.md#discardable-ratio) 和 [`max-background-gc`](/tikv-configuration-file.md#max-background-gc) 的设置对于 Titan 的读性能和垃圾回收过程都有重要影响。

当一个 blob file 中无用数据（相应的 key 已经被更新或删除）比例超过 [`discardable-ratio`](/tikv-configuration-file.md#discardable-ratio) 设置的阈值时，将会触发 Titan GC。减少这个阈值可以减少空间放大，但是会造成 Titan 更频繁 GC；增加这个值可以减少 Titan GC，减少相应的 I/O 带宽和 CPU 消耗，但是会增加磁盘空间占用。
```

> 当从 TiKV Details - Thread CPU - RocksDB CPU 监控中观察到 Titan GC 线程长期处于满负荷状态时，应该考虑调整  [`max-background-gc`](/tikv-configuration-file.md#max-background-gc) 增加 Titan GC 线程池大小。

Reviewer: ```suggestion
若果你通过**TiKV Details** - **Thread CPU** - **RocksDB CPU** 监控中观察到 Titan GC 线程长期处于满负荷状态时，应该考虑调整 [`max-background-gc`](/tikv-configuration-file.md#max-background-gc) 增加 Titan GC 线程池大小。
```

> 通过调整 [`rate-bytes-per-sec`](/tikv-configuration-file.md#rate-bytes-per-sec)，你可以限制 RocksDB compaction 的 I/O 速率，从而在高流量时减少对前台读写性能的影响。

Reviewer: ```suggestion
### `rate-bytes-per-sec`

通过调整 [`rate-bytes-per-sec`](/tikv-configuration-file.md#rate-bytes-per-sec)，你可以限制 RocksDB compaction 的 I/O 速率，从而在高流量时减少对前台读写性能的影响。
```

开启 Titan 以后，原有的数据并不会马上迁移到 Titan 引擎，而是随着前台写入和 RocksDB Compaction 的进行，**逐步进行 key-value 分离并写入 Titan**。同样的，无论是通过 [BR](/br/backup-and-restore-overview.md) 快照或日志恢复的数据或扩缩容过程中产生的数据搬迁，还是通过[TiDB Lightning](/tidb-lightning/tidb-lightning-overview.md) 物理导入模式导入的数据，都会首先写入 RocksDB。然后，随着 RocksDB Compaction 的进行，超过 [`min-blob-size`](/tikv-configuration-file.md#min-blob-size) 默认值 `32KB` 的大 value 会逐步分离到 Titan 中。你可以通过观察 **TiKV Details** - **Titan kv** - **blob file size** 监控面板中文件的大小来确认存储在 Titan 中的数据大小。

为了更快地将数据转移到 Titan，建议使用 tikv-ctl 工具执行一次全量 Compaction，以提高迁移速度。具体操作步骤请参考[手动 compact](/tikv-control.md#手动-compact-整个-tikv-集群的数据)。由于 RocksDB 具备 Block Cache，并且在将数据从 RocksDB 迁移到 Titan 时，数据访问是连续的，这使得在迁移过程中 Block Cache 能够更有效地提升迁移速度。在我们的测试中，通过 tikv-ctl 在单个 TiKV 节点上执行全量 Compaction，仅需 1 小时就能将 670 GiB 的数据迁移到 Titan。

需要注意的是，由于 Titan Blob 文件中的 Value 并非连续的，而且 Titan 的缓存是基于 Value 级别的，因此 Blob Cache 无法在 Compaction 过程中提供帮助。相较于从 RocksDB 转向 Titan 的速度，从 Titan 转回 RocksDB 的速度则会慢一个数量级。在测试中，通过 tikv-ctl 将 TiKV 节点上的 800 GiB Titan 数据进行全量 Compaction 转为 RocksDB，需要花费 12 个小时。

## 常用配置参数

> 通过合理配置 Titan 参数，可以有效提升数据库性能和资源利用率。

Reviewer: ```suggestion
通过合理配置 Titan 参数，可以有效提升数据库性能和资源利用率。本节介绍了一些常见的参数。
```

### `min-blob-size`

> 你可以通过设置 [`min-blob-size`](/tikv-configuration-file.md#min-blob-size) 来调整 value 的大小阈值，决定哪些数据保存在 RocksDB 中，哪些数据保存在 Titan 的 blob file 中。经过[测试](/storage-engine/titan-overview.md#min-blob-size-的选择及其性能影响)，`32KB` 是个折中的值。如果你想进一步提升写性能，并能接受扫描性能的下降，你可以将该值调整为 `1KB`。

Reviewer: ```suggestion
你可以通过设置 [`min-blob-size`](/tikv-configuration-file.md#min-blob-size) 来调整 value 的大小阈值，决定哪些数据保存在 RocksDB 中，哪些数据保存在 Titan 的 blob file 中。经过测试，`32KB` 是个折中的值。如果你想进一步提升写性能，并能接受扫描性能的下降，你可以将该值调整为 `1KB`。
```

### `blob-cache-size`

可以使用 [`blob-cache-size`](/tikv-configuration-file.md#blob-cache-size) 控制 Titan 中 value 的缓存大小。更大的缓存能提高 Titan 读性能，但过大的缓存会造成 OOM。

建议在数据库稳定运行后，根据监控把 RocksDB block cache (`storage.block-cache.capacity`) 设置为 store size 减去 blob file size 的大小，`blob-cache-size` 设置为内存大小 * 50% 减去 block cache 的大小。这是为了保证 block cache 足够缓存整个 RocksDB 的前提下，blob cache 尽量大。

### `discardable-ratio` 和 `max-background-gc`

[`discardable-ratio`](/tikv-configuration-file.md#discardable-ratio) 和 [`max-background-gc`](/tikv-configuration-file.md#max-background-gc) 的设置对于 Titan 的读性能和垃圾回收过程都有重要影响。

当一个 blob file 中无用数据（相应的 key 已经被更新或删除）比例超过 [`discardable-ratio`](/tikv-configuration-file.md#discardable-ratio) 设置的阈值时，将会触发 Titan GC。减少这个阈值可以减少空间放大，但是会造成 Titan 更频繁 GC；增加这个值可以减少 Titan GC，减少相应的 I/O 带宽和 CPU 消耗，但是会增加磁盘空间占用。

若果你通过**TiKV Details** - **Thread CPU** - **RocksDB CPU** 监控中观察到 Titan GC 线程长期处于满负荷状态时，应该考虑调整 [`max-background-gc`](/tikv-configuration-file.md#max-background-gc) 增加 Titan GC 线程池大小。

### `rate-bytes-per-sec`

通过调整 [`rate-bytes-per-sec`](/tikv-configuration-file.md#rate-bytes-per-sec)，你可以限制 RocksDB compaction 的 I/O 速率，从而在高流量时减少对前台读写性能的影响。

### Titan 配置文件示例

> 下面是一个 Titan 配置文件的样例，更多的参数说明，请参考 [TiKV 配置文件描述](/tikv-configuration-file.md)。你可以使用 TiUP [修改配置参数](/maintain-tidb-using-tiup.md#修改配置参数)，也可以通过[在 Kubernetes 中配置 TiDB 集群](https://docs.pingcap.com/zh/tidb-in-kubernetes/stable/configure-a-tidb-cluster) 修改配置参数。

Reviewer 3: ```suggestion
下面是一个 Titan 配置文件的样例，更多的参数说明，请参考 [TiKV 配置文件描述](/tikv-configuration-file.md)。你可以使用 TiUP [修改配置参数](/maintain-tidb-using-tiup.md#修改配置参数)，也可以通过[在 Kubernetes 中配置 TiDB 集群](https://docs.pingcap.com/zh/tidb-in-kubernetes/stable/configure-a-tidb-cluster)修改配置参数。
```

## 开启 Titan

> > **注意**

Reviewer: ```suggestion
> **注意：**
```

>

> > - 从 TiDB v7.6.0 开始，新集群将默认启用 Titan，并将阈值[`min-blob-size`](/tikv-configuration-file.md#min-blob-size)的默认值从之前版本的`1KB` 调整为`32KB`。

Reviewer: ```suggestion
> - 从 TiDB v7.6.0 开始，新集群将默认启用 Titan，并将阈值 [`min-blob-size`](/tikv-configuration-file.md#min-blob-size)的默认值由之前的 `1KB` 调整为 `32KB`。
```

> - 如果集群在升级到 TiDB v7.6.0 或更高版本之前未启用 Titan，则升级后将保持原有配置，继续使用 RocksDB。

> > - 如果集群在升级到 TiDB v7.6.0 或更高版本之前已经启用了 Titan，则升级后将维持原有配置，保持启用 Titan，并保留升级前[`min-blob-size`](/tikv-configuration-file.md#min-blob-size)的配置。如果升级前没有显式配置该值，则升级后仍然保持了老版本默认值`1KB`，以确保升级后集群配置的稳定性。

Reviewer: ```suggestion
> - 如果集群在升级到 TiDB v7.6.0 或更高版本之前已经启用了 Titan，则升级后将维持原有配置，保持启用 Titan，并保留升级前 [`min-blob-size`](/tikv-configuration-file.md#min-blob-size) 的配置。如果升级前没有显式配置该值，则升级后仍然保持老版本的默认值 `1KB`，以确保升级后集群配置的稳定性。
```

+ 方法二：直接编辑 TiKV 配置文件开启 Titan（生产环境不推荐）。

    ``` toml
    [rocksdb.titan]
    enabled = true
    ```

+ 方法三：编辑 TiDB-Operator 的 `${cluster_name}/tidb-cluster.yaml` 配置文件，编辑示例如下：

>     ``` yaml

Reviewer: ```suggestion
    ```yaml
```

### `discardable-ratio` 和 `max-background-gc`

[`discardable-ratio`](/tikv-configuration-file.md#discardable-ratio) 和 [`max-background-gc`](/tikv-configuration-file.md#max-background-gc) 的设置对于 Titan 的读性能和垃圾回收过程都有重要影响。当一个 blob file 中无用数据（相应的 key 已经被更新或删除）比例超过 [`discardable-ratio`](/tikv-configuration-file.md#discardable-ratio) 设置的阈值时，将会触发 Titan GC。减少这个阈值可以减少空间放大，但是会造成 Titan 更频繁 GC；增加这个值可以减少 Titan GC，减少相应的 I/O 带宽和 CPU 消耗，但是会增加磁盘空间占用。

当从 TiKV Details - Thread CPU - RocksDB CPU 监控中观察到 Titan GC 线程长期处于满负荷状态时，应该考虑调整  [`max-background-gc`](/tikv-configuration-file.md#max-background-gc) 增加 Titan GC 线程池大小。

通过调整 [`rate-bytes-per-sec`](/tikv-configuration-file.md#rate-bytes-per-sec)，你可以限制 RocksDB compaction 的 I/O 速率，从而在高流量时减少对前台读写性能的影响。

### Titan 配置文件示例

下面是一个 Titan 配置文件的样例，更多的参数说明，请参考 [TiKV 配置文件描述](/tikv-configuration-file.md)。你可以使用 TiUP [修改配置参数](/maintain-tidb-using-tiup.md#修改配置参数)，也可以通过[在 Kubernetes 中配置 TiDB 集群](https://docs.pingcap.com/zh/tidb-in-kubernetes/stable/configure-a-tidb-cluster) 修改配置参数。

> ``` toml

Reviewer 4: ```suggestion
```toml
```

+ 方法二：直接编辑 TiKV 配置文件开启 Titan（生产环境不推荐）。

    ``` toml
    [rocksdb.titan]
    enabled = true
    ```

+ 方法三：编辑 TiDB Operator 的 `${cluster_name}/tidb-cluster.yaml` 配置文件，编辑示例如下：

    ```yaml
    spec:
      tikv:
        ## Base image of the component
        baseImage: pingcap/tikv
        ## tikv-server configuration
        ## Ref: https://docs.pingcap.com/zh/tidb/stable/tikv-configuration-file
        config: |
          log-level = "info"
          [rocksdb]
            [rocksdb.titan]
              enabled = true
    ```

    应用配置时，触发在线滚动重启 TiDB 集群让配置生效：

    ```shell
    kubectl apply -f ${cluster_name} -n ${namespace}
    ```

    更多信息请参考[在 Kubernetes 中配置 TiDB 集群](https://docs.pingcap.com/zh/tidb-in-kubernetes/stable/configure-a-tidb-cluster)。

## 数据迁移

> **警告：**
>
> 在关闭 Titan 功能的情况下，RocksDB 无法读取已经迁移到 Titan 的数据。如果在打开过 Titan 的 TiKV 实例上错误地关闭了 Titan（误设置 `rocksdb.titan.enabled = false`），启动 TiKV 会失败，TiKV log 中出现 `You have disabled titan when its data directory is not empty` 的错误。如需要关闭 Titan，请参考[关闭 Titan](#关闭-titan) 。

> 开启 Titan 以后，原有的数据并不会马上迁移到 Titan 引擎，而是随着前台写入和 RocksDB Compaction 的进行，**逐步进行 key-value 分离并写入 Titan**。同样的，无论是通过 [BR](/br/backup-and-restore-overview.md) 快照或日志恢复的数据，还是通过 [TiDB Lightning](/tidb-lightning/tidb-lightning-overview.md) 逻辑导入模式或物理导入模式导入的数据，都会首先写入 RocksDB。然后，随着 RocksDB Compaction 的进行，超过 [`min-blob-size`](/tikv-configuration-file.md#min-blob-size) 默认值 `32KB` 的大 value 会逐步分离到 Titan 中。你可以通过观察 **TiKV Details** - **Titan kv** - **blob file size** 监控面板中文件的大小来确认存储在 Titan 中的数据大小。

Reviewer 5: 是不是也要说下 region snapshot 也是只有 titan，所以对于新加的节点，搬过去的数据都不在 titan 中

    enabled = true
    ```

> 开启 Titan 以后，原有的数据并不会马上移入 Titan 引擎，而是随着前台写入和 RocksDB compaction 的进行，逐步进行 key-value 分离并写入 Titan。同样，全量、增量恢复或者Lightning导入的SST都是RocksDB格式，数据不会直接导入Titan。在Compaction过程中做数据搬迁到Titan。可以通过观察 **TiKV Details** - **Titan kv** - **blob file size** 监控面版确认数据保存在 Titan 中部分的大小。

Reviewer 5: ```suggestion
开启 Titan 以后，原有的数据并不会马上移入 Titan 引擎，而是随着前台写入和 RocksDB compaction 的进行，逐步进行 key-value 分离并写入 Titan。同样，全量、增量恢复或者Lightning导入的SST都是RocksDB格式，数据不会直接导入Titan。随着Compaction的进行，大 value 会慢慢的分离到 Titan 中。可以通过观察 **TiKV Details** - **Titan kv** - **blob file size** 监控面版确认数据保存在 Titan 中部分的大小。
```

    enabled = true
    ```

> 

Reviewer 6: 由于 Titan 默认开启了，建议在这里增加上如果通过 TiDB-Operator 开启/关闭 Titan 的说明

Author: @benmaoer 这个和TiDB-Operator有什么关系呢，如果要增加tidb-operator的说明，可以在另外的PR里。另外，我也不知道tidb-operator的用法

> 

Reviewer 6: ```suggestion

+ 方法三：编辑 TiDB-Operator 的 ${cluster_name}/tidb-cluster.yaml 配置文件，编辑示例如下：
  
    {{< copyable "" >}}

    ``` yaml
      tikv:
        ## Base image of the component
        baseImage: pingcap/tikv

        ## tikv-server configuration
        ## Ref: https://docs.pingcap.com/tidb/stable/tikv-configuration-file
        config: |
          log-level = "info"      

          [rocksdb]
            [rocksdb.titan]
              enabled = true
    ```

    应用配置时默认会触发在线滚动重启 TiDB 集群让配置生效：

     {{< copyable "shell-regular" >}}

    ```shell
    kubectl apply -f ${cluster_name} -n ${namespace}
    ```   

    更多信息可参考[在 Kubernetes 中配置 TiDB 集群](https://docs.pingcap.com/zh/tidb-in-kubernetes/stable/configure-a-tidb-cluster)。

```

> 开启 Titan 以后，原有的数据并不会马上移入 Titan 引擎，而是随着前台写入和 RocksDB compaction 的进行，逐步进行 key-value 分离并写入 Titan。同样，全量、增量恢复或者 TiDB Lightning 导入的 SST 都是 RocksDB 格式，数据不会直接导入Titan。随着 Compaction 的进行，被处理过的 SST 中的大 value 会分离到 Titan 中。可以通过观察 **TiKV Details** - **Titan kv** - **blob file size** 监控面版确认数据保存在 Titan 中部分的大小。

Reviewer: ```suggestion
开启 Titan 以后，原有的数据并不会马上移入 Titan 引擎，而是随着前台写入和 RocksDB compaction 的进行，逐步进行 key-value 分离并写入 Titan。同样，全量、增量恢复或者 TiDB Lightning 导入的 SST 都是 RocksDB 格式，数据不会直接导入 Titan。随着 Compaction 的进行，被处理过的 SST 中的大 value 会分离到 Titan 中。可以通过观察 **TiKV Details** - **Titan kv** - **blob file size** 监控面版确认数据保存在 Titan 中部分的大小。
```

> 开启 Titan 以后，原有的数据并不会马上移入 Titan 引擎，而是随着前台写入和 RocksDB compaction 的进行，逐步进行 key-value 分离并写入 Titan。同样，全量、增量恢复或者 TiDB Lightning 导入的 SST 都是 RocksDB 格式，数据不会直接导入 Titan。随着 Compaction 的进行，被处理过的 SST 中的大 value 会分离到 Titan 中。可以通过观察 **TiKV Details** - **Titan kv** - **blob file size** 监控面版确认数据保存在 Titan 中部分的大小。

Reviewer 6: ```suggestion
开启 Titan 以后，原有的数据并不会马上迁移到 Titan 引擎，而是随着前台写入和 RocksDB Compaction 的进行，逐步进行 key-value 分离并写入 Titan。同样的，无论是通过 [BR 快照/日志恢复](/br/backup-and-restore-overview.md) 的数据，还是通过 [TiDB Lightning](/tidb-lightning/tidb-lightning-overview.md) 逻辑/物理导入的数据，都是先写入 RocksDB，再通过 RocksDB Compaction 的进行，逐步分离超过 [32KB](tikv-configuration-file.md#min-blob-size) 的大 value 到 Titan 中。用户可以通过观察 **TiKV Details** - **Titan kv** - **blob file size** 监控面版中文件的大小来确认存储在 Titan 中的数据大小。
```

> 如果需要加速数据移入 Titan，可以通过 tikv-ctl 执行一次全量 compaction，请参考[手动 compact](/tikv-control.md#手动-compact-整个-tikv-集群的数据)。由于 RocksDB 有 Block Cache，且转成 Titan 时的数据访问是连续的，因此 Block Cache 能有很好的命中率。在测试中，TiKV 节点上 670 GiB 的数据，通过 tikv-ctl 全量 compaction 转成 Titan，只需 1 小时。

Reviewer 6: ```suggestion
如果希望加快数据迁移到 Titan 的过程，可以通过 tikv-ctl 执行一次全量 Compaction，请参考[手动 compact](/tikv-control.md#手动-compact-整个-tikv-集群的数据)。由于 RocksDB 有 Block Cache，且转成 Titan 时的数据访问是连续的，因此 Block Cache 能有很好的命中率。在测试中，仅需要 1 个小时即可通过 tikv-ctl 完成单个TiKV 节点上 670 GiB 的数据的全量 Compaction。
```

> 如果需要加速数据移入 Titan，可以通过 tikv-ctl 执行一次全量 compaction，请参考[手动 compact](/tikv-control.md#手动-compact-整个-tikv-集群的数据)。由于 RocksDB 有 Block Cache，且转成 Titan 时的数据访问是连续的，因此Block Cache 能有很好的命中率。在测试中，一个 670 GiB 的 TiKV 节点数据，通过 tikv-ctl 全量 compaction 转成 Titan，只需 1 小时。

Reviewer: ```suggestion
如果需要加速数据移入 Titan，可以通过 tikv-ctl 执行一次全量 compaction，请参考[手动 compact](/tikv-control.md#手动-compact-整个-tikv-集群的数据)。由于 RocksDB 有 Block Cache，且转成 Titan 时的数据访问是连续的，因此 Block Cache 能有很好的命中率。在测试中，一个 670 GiB 的 TiKV 节点数据，通过 tikv-ctl 全量 compaction 转成 Titan，只需 1 小时。
```

> 如果需要加速数据移入 Titan，可以通过 tikv-ctl 执行一次全量 compaction，请参考[手动 compact](/tikv-control.md#手动-compact-整个-tikv-集群的数据)。由于 RocksDB 有 Block Cache，且转成 Titan 时的数据访问是连续的，因此Block Cache 能有很好的命中率。在测试中，一个 670 GiB 的 TiKV 节点数据，通过 tikv-ctl 全量 compaction 转成 Titan，只需 1 小时。
> 
> > **注意：**
> >
> > 在7.6.0开始，新建集群默认打开Titan，已有集群升级到7.6.0则会维持原有的配置, 没有显式配置titan的，则仍然会使用RocksDB。

Reviewer: ```suggestion
如果需要加速数据移入 Titan，可以通过 tikv-ctl 执行一次全量 compaction，请参考[手动 compact](/tikv-control.md#手动-compact-整个-tikv-集群的数据)。由于 RocksDB 有 Block Cache，且转成 Titan 时的数据访问是连续的，因此Block Cache 能有很好的命中率。在测试中，一个 670 GiB 的 TiKV 节点数据，通过 tikv-ctl 全量 compaction 转成 Titan，只需 1 小时。

> **注意：**
>
> 在 v7.6.0 开始，新建集群默认打开 Titan。已有集群升级到 v7.6.0 则会维持原有的配置，如果没有显式配置 Titan，则仍然会使用 RocksDB。
```

> > 在 v7.6.0 开始，新建集群默认打开 Titan。已有集群升级到 v7.6.0 则会维持原有的配置，如果没有显式配置 Titan，则仍然会使用 RocksDB。

Reviewer 6: ```suggestion
> - 从 TiDB v7.6.0 开始，新创建的集群将默认启用 Titan。写入的数据中，超过 [`min-blob-size`](tikv-configuration-file.md#min-blob-size)  的 value 将被存储在 Titan 中，而其他数据则继续存储在 RocksDB 中。
> - 对于已存在的集群，如果在升级到 TiDB v7.6.0 或更高版本之前未启用 Titan，则升级后将保持原有配置，继续使用 RocksDB，不会启用 Titan。
> - 对于已存在的集群，如果在升级到 TiDB v7.6.0 或更高版本之前已经启用了 Titan，则升级后将维持原有配置，保持启用 Titan 引擎，并保留升级前 [`min-blob-size`](tikv-configuration-file.md#min-blob-size)  的配置，可以是原默认 1KB ，也可以是用户定义的其他值。
```

> > 在 v7.6.0 开始，新建集群默认打开 Titan。已有集群升级到 v7.6.0 则会维持原有的配置，如果没有显式配置 Titan，则仍然会使用 RocksDB。

Reviewer: ```suggestion
> - 从 v7.6.0 开始，新建集群默认打开 Titan。已有集群升级到 v7.6.0 则会维持原有的配置，如果没有显式配置 Titan，则仍然会使用 RocksDB。
 > - 从 v7.6.0 开始，新建集群的 `min-blob-size` 默认值由 `1KB` 改为 `32KB`，避免扫描性能产生回退。但已有集群升级到 v7.6.0，则会维持原有的配置，即如果没有显式配置 `min-blob-size`，则保持旧版本的默认值 `1KB` 不变。
```

    enabled = true
    ```

+ 方法三：编辑 TiDB-Operator 的 `${cluster_name}/tidb-cluster.yaml` 配置文件，编辑示例如下：

>      {{< copyable "" >}}
>  

Reviewer: ```suggestion
```

    ``` yaml
      tikv:
        ## Base image of the component
        baseImage: pingcap/tikv
        ## tikv-server configuration
        ## Ref: https://docs.pingcap.com/tidb/stable/tikv-configuration-file
        config: |
          log-level = "info"
          [rocksdb]
            [rocksdb.titan]
              enabled = true
    ```

>     应用配置时，触发在线滚动重启 TiDB 集群让配置生效：

Reviewer: ```suggestion

    应用配置时，触发在线滚动重启 TiDB 集群让配置生效：
```

    {{< copyable "" >}}

    ```shell
    kubectl apply -f ${cluster_name} -n ${namespace}
    ```
    更多信息可参考[在 Kubernetes 中配置 TiDB 集群](https://docs.pingcap.com/zh/tidb-in-kubernetes/stable/configure-a-tidb-cluster)。

## 数据迁移

> > **警告**

Reviewer: ```suggestion
> **警告：**
```

>

> > 在不开启 Titan 功能的情况下，RocksDB 无法读取已经迁移到 Titan 的数据。如果在打开过 Titan 的 TiKV 实例上错误地关闭了 Titan（误设置 `rocksdb.titan.enabled = false`），启动 TiKV 会失败，TiKV log 中出现 `You have disabled titan when its data directory is not empty` 错误。如需要关闭 Titan，参考[关闭 Titan](#关闭-titan) 一节。

Reviewer: ```suggestion
> 在关闭 Titan 功能的情况下，RocksDB 无法读取已经迁移到 Titan 的数据。如果在打开过 Titan 的 TiKV 实例上错误地关闭了 Titan（误设置 `rocksdb.titan.enabled = false`），启动 TiKV 会失败，TiKV log 中出现 `You have disabled titan when its data directory is not empty` 的错误。如需要关闭 Titan，参考[关闭 Titan](#关闭-titan) 。
```

> 开启 Titan 以后，原有的数据并不会马上迁移到 Titan 引擎，而是随着前台写入和 RocksDB Compaction 的进行，**逐步进行 key-value 分离并写入 Titan**。同样的，无论是通过 [BR](/br/backup-and-restore-overview.md) 快照/日志恢复的数据，还是通过 [TiDB Lightning](/tidb-lightning/tidb-lightning-overview.md) 逻辑/物理导入的数据，都会首先写入 RocksDB。然后，随着 RocksDB Compaction 的进行，超过 [`min-blob-size`](/tikv-configuration-file.md#min-blob-size) 默认值 32KB 的大 value 会逐步分离到 Titan 中。用户可以通过观察 **TiKV Details** - **Titan kv** - **blob file size** 监控面版中文件的大小来确认存储在 Titan 中的数据大小。

Reviewer: ```suggestion
开启 Titan 以后，原有的数据并不会马上迁移到 Titan 引擎，而是随着前台写入和 RocksDB Compaction 的进行，**逐步进行 key-value 分离并写入 Titan**。同样的，无论是通过 [BR](/br/backup-and-restore-overview.md) 快照或日志恢复的数据，还是通过 [TiDB Lightning](/tidb-lightning/tidb-lightning-overview.md) 逻辑导入模式或物理导入模式导入的数据，都会首先写入 RocksDB。然后，随着 RocksDB Compaction 的进行，超过 [`min-blob-size`](/tikv-configuration-file.md#min-blob-size) 默认值 `32KB` 的大 value 会逐步分离到 Titan 中。你可以通过观察 **TiKV Details** - **Titan kv** - **blob file size** 监控面版中文件的大小来确认存储在 Titan 中的数据大小。
```

> 为了更快地将数据转移到 Titan，建议使用 tikv-ctl 工具执行一次全量 Compaction，以提高迁移速度。具体的操作步骤可以查阅[手动 compact](/tikv-control.md#手动-compact-整个-tikv-集群的数据)。由于 RocksDB 具备 Block Cache，并且在将数据从 RocksDB 迁移到 Titan 时，数据访问是连续的，这使得在迁移过程中 Block Cache 能够更有效地提升迁移速度。在我们的测试中，仅需 1 小时，就能够通过 tikv-ctl 在单个 TiKV 节点上执行全量 Compaction，成功将 670 GiB 的数据迁移到 Titan。

Reviewer: ```suggestion
为了更快地将数据转移到 Titan，建议使用 tikv-ctl 工具执行一次全量 Compaction，以提高迁移速度。具体操作步骤请参考[手动 compact](/tikv-control.md#手动-compact-整个-tikv-集群的数据)。由于 RocksDB 具备 Block Cache，并且在将数据从 RocksDB 迁移到 Titan 时，数据访问是连续的，这使得在迁移过程中 Block Cache 能够更有效地提升迁移速度。在我们的测试中，仅需 1 小时，就能够通过 tikv-ctl 在单个 TiKV 节点上执行全量 Compaction，成功将 670 GiB 的数据迁移到 Titan。
```

需要注意的是，由于 Titan Blob 文件中的 Value 并非连续的，而且 Titan 的缓存是基于 Value 级别的，因此 Blob Cache 无法在 Compaction 过程中提供帮助。相较于从 RocksDB 转向 Titan 的速度，从 Titan 转回 RocksDB 的速度则会慢一个数量级。在测试中，通过 tikv-ctl 将 TiKV 节点上的 800 GiB Titan 数据进行全量 Compaction 转为 RocksDB，需要花费 12 个小时。

## 常用配置

通过合理配置 Titan 参数，可以有效提升数据库性能和资源利用率。在 Titan 的配置中，我们可以通过设置[`min-blob-size`](/tikv-configuration-file.md#min-blob-size)来调整 value 的大小阈值，决定哪些数据保存在 RocksDB 中，哪些数据保存在 Titan 的 blob file 中。经过[测试](/storage-engine/titan-overview.md#min-blob-size-的选择及其性能影响)，`32KB` 是个折中的值。

此外，[`blob-file-compression`](/tikv-configuration-file.md#blob-file-compression) 参数可以指定 Titan 中 value 所使用的压缩算法，而 [`zstd-dict-size`](/tikv-configuration-file.md#zstd-dict-size) 可以通过启用 zstd 字典压缩来提高压缩率。

要注意的是，[`blob-cache-size`](/tikv-configuration-file.md#blob-cache-size) 控制了Titan 中 value 的缓存大小。更大的缓存能提高 Titan 读性能，但过大的缓存会造成 OOM。建议在数据库稳定运行后，根据监控把 RocksDB block cache (storage.block-cache.capacity) 设置为 store size 减去 blob file size 的大小，blob-cache-size 设置为 内存大小 * 50% 再减去 block cache 的大小。这是为了保证 block cache 足够缓存整个 RocksDB 的前提下，blob cache 尽量大。

> [`discardable-ratio`](/tikv-configuration-file.md#discardable-ratio)和[`max-background-gc`](/tikv-configuration-file.md#max-background-gc)的设置对于 Titan 的读性能和垃圾回收过程都有重要影响。当一个 blob file 中无用数据（相应的 key 已经被更新或删除）比例超过[`discardable-ratio`](/tikv-configuration-file.md#discardable-ratio)设置的阈值时，将会触发 Titan GC。减少这个阈值可以减少空间放大，但是会造成 Titan 更频繁 GC；增加这个值可以减少 Titan GC，减少相应的 I/O 带宽和 CPU 消耗，但是会增加磁盘空间占用。当从 TiKV Details - Thread CPU - RocksDB CPU 监控中观察到 Titan GC 线程长期处于满负荷状态时，应该考虑调整 [`max-background-gc`](/tikv-configuration-file.md#max-background-gc) 增加 Titan GC 线程池大小。

Reviewer: ```suggestion
[`discardable-ratio`](/tikv-configuration-file.md#discardable-ratio) 和 [`max-background-gc`](/tikv-configuration-file.md#max-background-gc) 的设置对于 Titan 的读性能和垃圾回收过程都有重要影响。当一个 blob file 中无用数据（相应的 key 已经被更新或删除）比例超过 [`discardable-ratio`](/tikv-configuration-file.md#discardable-ratio) 设置的阈值时，将会触发 Titan GC。减少这个阈值可以减少空间放大，但是会造成 Titan 更频繁 GC；增加这个值可以减少 Titan GC，减少相应的 I/O 带宽和 CPU 消耗，但是会增加磁盘空间占用。当从 TiKV Details - Thread CPU - RocksDB CPU 监控中观察到 Titan GC 线程长期处于满负荷状态时，应该考虑调整  [`max-background-gc`](/tikv-configuration-file.md#max-background-gc) 增加 Titan GC 线程池大小。
```

> 最后，通过调整[`rate-bytes-per-sec`](/tikv-configuration-file.md#rate-bytes-per-sec)，我们能够限制 RocksDB compaction 的 I/O 速率，从而在高流量时减少对前台读写性能的影响。 

Reviewer: ```suggestion
最后，通过调整 [`rate-bytes-per-sec`](/tikv-configuration-file.md#rate-bytes-per-sec)，可以限制 RocksDB compaction 的 I/O 速率，从而在高流量时减少对前台读写性能的影响。 
```

> 下面是一个 Titan 配置文件的样例，更多的参数说明，请参考[TiKV 配置文件描述](/tikv-configuration-file.md)。我们可以使用 TiUP 来[修改配置参数](/maintain-tidb-using-tiup.md#修改配置参数)，也可以通过 [在 Kubernetes 中配置 TiDB 集群](https://docs.pingcap.com/zh/tidb-in-kubernetes/stable/configure-a-tidb-cluster) 来修改配置参数。

Reviewer: ```suggestion
下面是一个 Titan 配置文件的样例，更多的参数说明，请参考 [TiKV 配置文件描述](/tikv-configuration-file.md)。你可以使用 TiUP [修改配置参数](/maintain-tidb-using-tiup.md#修改配置参数)，也可以通过[在 Kubernetes 中配置 TiDB 集群](https://docs.pingcap.com/zh/tidb-in-kubernetes/stable/configure-a-tidb-cluster) 修改配置参数。
```

>     {{< copyable "" >}}

Reviewer: ```suggestion
```

+ value 的大小阈值。

    当写入的 value 小于这个值时，value 会保存在 RocksDB 中，反之则保存在 Titan 的 blob file 中。根据 value 大小的分布，增大这个值可以使更多 value 保存在 RocksDB，读取这些小 value 的性能会稍好一些；减少这个值可以使更多 value 保存在 Titan 中，进一步减少 RocksDB compaction。经过[测试](/storage-engine/titan-overview.md#min-blob-size的选择及其性能影响)，1KB是一个比较折中的值。如果发现系统磁盘占用过大可以进一步调高这个值。

    ```toml
    [rocksdb.defaultcf.titan]
    min-blob-size = "1KB"
    ```

> + Titan 中 value 所使用的压缩算法。默认情况下，Titan 中压缩是以 value 为单元的。因此Titan的压缩率低于RocksDB相比。以Json内容为例，Titan的store size可能比RocksDB高30%至50%。 用户可以通过设置zstd_dict_size（比如16KB）启用字典ZSTD以大幅提高压缩率（实际Store Size可以低于RocksDB），但字典ZSTD在有些负载下会有10%左右的性能损失。

Author: 已添加的配置文档中。

+ value 的大小阈值。

>     当写入的 value 小于这个值时，value 会保存在 RocksDB 中，反之则保存在 Titan 的 blob file 中。根据 value 大小的分布，增大这个值可以使更多 value 保存在 RocksDB，读取这些小 value 的性能会稍好一些；减少这个值可以使更多 value 保存在 Titan 中，进一步减少 RocksDB compaction。经过[测试](/storage-engine/titan-overview.md#min-blob-size的选择及其性能影响)，1 KB是一个比较折中的值。如果发现系统磁盘占用过大可以进一步调高这个值。

Reviewer: ```suggestion
    当写入的 value 小于这个值时，value 会保存在 RocksDB 中，反之则保存在 Titan 的 blob file 中。根据 value 大小的分布，增大这个值可以使更多 value 保存在 RocksDB，读取这些小 value 的性能会稍好一些；减少这个值可以使更多 value 保存在 Titan 中，进一步减少 RocksDB compaction。经过[测试](/storage-engine/titan-overview.md#min-blob-size-的选择及其性能影响)，1 KB是一个比较折中的值。如果发现系统磁盘占用过大可以进一步调高这个值。
```

    ```toml
    [rocksdb.defaultcf.titan]
    min-blob-size = "1KB"
    ```

> + Titan 中 value 所使用的压缩算法。在7.6版本开始，默认采用zstd压缩算法。

Reviewer: ```suggestion
+ Titan 中 value 所使用的压缩算法。从 v7.6.0 开始，默认采用 `zstd` 压缩算法。
```

    ```toml
    [rocksdb.defaultcf.titan]
    blob-file-compression = "zstd"
    ```

> + 默认情况下, zstd-dict-size为 0KB , 表示Titan中压缩的是单个value值，而RocksDB压缩以Block(默认 32KB )为单位。因此当value平均小于 32KB 时，Titan的压缩率低于RocksDB。以Json内容为例，Titan的store size可能比RocksDB高30%至50%。实际压缩率还取决于value内容是否适合压缩，以及不同value之间的相似性。用户可以通过设置zstd-dict-size（比如 16KB ）启用zstd字典压缩以大幅提高压缩率（实际Store Size可以低于RocksDB），但zstd字典压缩在有些负载下会有10%左右的性能损失。

Reviewer: ```suggestion
+ 默认情况下，`zstd-dict-size` 为 `0KB` , 表示 Titan 中压缩的是单个 value 值，而 RocksDB 压缩以 Block（默认值为 `32KB`）为单位。因此当 value 平均小于 32KB 时，Titan 的压缩率低于 RocksDB。以 JSON 内容为例，Titan 的 store size 可能比 RocksDB 高 30% 至 50%。实际压缩率还取决于 value 内容是否适合压缩，以及不同 value 之间的相似性。你可以通过设置 `zstd-dict-size`（比如 `16KB` ）启用 `zstd` 字典压缩以大幅提高压缩率（实际 Store Size 可以低于 RocksDB）。但 `zstd` 字典压缩在有些负载下会有 10% 左右的性能损失。
```

> + 默认情况下，`zstd-dict-size` 为 `0KB` , 表示 Titan 中压缩的是单个 value 值，而 RocksDB 压缩以 Block（默认值为 `32KB`）为单位。因此当 value 平均小于 32KB 时，Titan 的压缩率低于 RocksDB。以 JSON 内容为例，Titan 的 store size 可能比 RocksDB 高 30% 至 50%。实际压缩率还取决于 value 内容是否适合压缩，以及不同 value 之间的相似性。你可以通过设置 `zstd-dict-size`（比如 `16KB` ）启用 ZSTD 字典压缩以大幅提高压缩率（实际 Store Size 可以低于 RocksDB）。但  ZSTD 字典压缩在有些负载下会有 10% 左右的性能损失。

Reviewer: ```suggestion
+ 默认情况下，`zstd-dict-size` 为 `0KB`，表示 Titan 中压缩的是单个 value 值，而 RocksDB 压缩以 Block（默认值为 `32KB`）为单位。因此当 value 平均小于 32KB 时，Titan 的压缩率低于 RocksDB。以 JSON 内容为例，Titan 的 store size 可能比 RocksDB 高 30% 至 50%。实际压缩率还取决于 value 内容是否适合压缩，以及不同 value 之间的相似性。你可以通过设置 `zstd-dict-size`（比如 `16KB` ）启用 ZSTD 字典压缩以大幅提高压缩率（实际 Store Size 可以低于 RocksDB）。但  ZSTD 字典压缩在有些负载下会有 10% 左右的性能损失。
```

+ value 的大小阈值。

>     当写入的 value 小于这个值时，value 会保存在 RocksDB 中，反之则保存在 Titan 的 blob file 中。根据 value 大小的分布，增大这个值可以使更多 value 保存在 RocksDB，读取这些小 value 的性能会稍好一些；减少这个值可以使更多 value 保存在 Titan 中，进一步减少 RocksDB compaction。经过[测试](/storage-engine/titan-overview.md#min-blob-size-的选择及其性能影响)，1 KB 是一个比较折中的值。如果发现系统磁盘占用过大，可以进一步调高这个值。

Reviewer 6: @tonyxuqqi @Connor1996 这里我们还推荐 1KB 的 min-blob-size 么？

    ```toml
    [rocksdb.defaultcf.titan]
    min-blob-size = "32KB"
    ```

> **注意：**
>

> > 在 v7.6.0 开始，新建集群默认min-blob-size改成了`32KB`, 避免扫描性能产生回退。但对已有集群升级到 v7.6.0 则会维持原有的配置，即便没有显式配置min-blob-size，仍然会使用`1KB`的旧默认值。

Reviewer 6: 建议和之前的升级默认开启 Titan 的说明放在一起

Author: 嗯，上面已经集中说明了。

> > **注意：**
> >
> > 在 v7.6.0 开始，新建集群默认min-blob-size改成了`32KB`, 避免扫描性能产生回退。但对已有集群升级到 v7.6.0 则会维持原有的配置，即便没有显式配置min-blob-size，仍然会使用`1KB`的旧默认值。

Reviewer: ```suggestion
```

    min-blob-size = "1KB"
    ```

> + Titan 中 value 所使用的压缩算法。默认情况下，Titan 中压缩是以 value 为单元的。因此Titan的压缩率低于RocksDB相比。以Json内容为例，Titan的store size可能比RocksDB高30%至50%。 用户可以通过设置zstd_dict_size（比如16KB）启用字典ZSTD以大幅提高压缩率（实际Store Size可以低于RocksDB），但字典ZSTD在有些负载下会有10%左右的性能损失。

Reviewer 2: 这个是不是要把 json 和完全随机的情况下的压缩率都说明一下？

Author: 完全随机的Value在实际系统里很少出现，而且如果是完全随机的话titan和rocksdb也没区别，用户也就不用关心压缩算法了。

Reviewer 2: 但是当时的确是出现了 rocksdb 数据低于 titan 几百个 GB，或者把这个现象给用户解释一下也好

Author: 我看看怎么写比较好。这个很难解释，随机其实也不是真的完全随机，每个value之间还是有相似性的，不然rocksdb也没法压缩。
之前测试出现这个问题的情况是每个value本身不大，也内部缺乏相似性，但是不同value之间有一些相似点，放一起也能压缩一些。

> - 当设置为 `read-only` 时，新写入的 value 不论大小均会写入 RocksDB。

Reviewer 2: https://docs.pingcap.com/zh/tidb/stable/tikv-configuration-file#blob-run-mode 这里配置是 read_only，需要统一一下

Author: 好的，应该是read-only，那里要改一下。

- 当设置为 `fallback` 时，新写入的 value 不论大小均会写入 RocksDB，并且当 RocksDB 进行 compaction 时，会自动把所碰到的存储在 Titan blob file 中的 value 移回 RocksDB。

> 如果现有数据和未来数据均不再需要 Titan，可执行以下步骤完全关闭 Titan。然而一般情况下只需要执行以下步骤1和步骤4即可，步骤2、3会影响用户SQL的性能。在Compaction过程中会将数据从Titan移回RocksDB。

Reviewer 2: 实测，没有步骤 2 对前台业务影响也较大

Author: @lilinghai 大概影响多大？主要是什么原因 @Connor1996

Reviewer 2: ![image](https://github.com/pingcap/docs-cn/assets/7677963/92895865-8fd7-4c62-bf20-ae8ff836a047)

- 当设置为 `read-only` 时，新写入的 value 不论大小均会写入 RocksDB。
- 当设置为 `fallback` 时，新写入的 value 不论大小均会写入 RocksDB，并且当 RocksDB 进行 compaction 时，会自动把所碰到的存储在 Titan blob file 中的 value 移回 RocksDB。

> 如果现有数据和未来数据均不再需要 Titan，可执行以下步骤完全关闭 Titan。然而一般情况下只需要执行以下步骤1和步骤3、4即可，步骤2会加快数据迁移速度，但会严重影响用户SQL的性能。事实上即便跳过步骤2，由于在Compaction过程中会将数据从Titan迁移到RocksDB，会占用额外的 IO 和 CPU 资源，因此仍然可以观察到一定的性能损失，在资源紧张的情况下吞吐可以下降 50% 以上。

Reviewer: ```suggestion
如果现有数据和未来数据均不再需要 Titan，可执行以下步骤完全关闭 Titan。然而一般情况下只需要执行以下步骤 1 和步骤 3、步骤 4 即可，步骤 2 会加快数据迁移速度，但会严重影响用户 SQL 的性能。事实上即便跳过步骤 2，由于在Compaction 过程中会将数据从 Titan 迁移到 RocksDB，会占用额外的 IO 和 CPU 资源，因此仍然可以观察到一定的性能损失，在资源紧张的情况下吞吐可以下降 50% 以上。
```

> 如果现有数据和未来数据均不再需要 Titan，可执行以下步骤完全关闭 Titan。然而一般情况下只需要执行以下步骤 1 和步骤 3、步骤 4 即可，步骤 2 会加快数据迁移速度，但会严重影响用户 SQL 的性能。事实上即便跳过步骤 2，由于在Compaction 过程中会将数据从 Titan 迁移到 RocksDB，会占用额外的 IO 和 CPU 资源，因此仍然可以观察到一定的性能损失，在资源紧张的情况下吞吐可以下降 50% 以上。

Reviewer: ```suggestion
如果现有数据和未来数据均不再需要 Titan，可执行以下步骤完全关闭 Titan。然而一般情况下只需要执行以下步骤 1 和步骤 3、步骤 4 即可，步骤 2 会加快数据迁移速度，但会严重影响用户 SQL 的性能。事实上即便跳过步骤 2，由于在 Compaction 过程中会将数据从 Titan 迁移到 RocksDB，会占用额外的 IO 和 CPU 资源，因此仍然可以观察到一定的性能损失，在资源紧张的情况下吞吐可以下降 50% 以上。
```

    discardable-ratio = 1.0
    ```

2. [可选] 使用 `tikv-ctl` 执行全量数据整理 (Compaction)。这一步骤将消耗大量 I/O 和 CPU 资源。

    ```bash
    tikv-ctl --pd <PD_ADDR> compact-cluster --bottommost force
    ```

> 3. [可选]数据整理结束后，通过 **TiKV-Details**/**Titan - kv** 监控面板确认 **Blob file count** 指标降为 0。

Reviewer 5: ```suggestion
3. 数据整理结束后，通过 **TiKV-Details**/**Titan - kv** 监控面板确认 **Blob file count** 指标降为 0。
```

    discardable-ratio = 1.0
    ```

>     > **注意**

Reviewer: ```suggestion
    > **注意：**
```

    discardable-ratio = 1.0
    ```

    > **注意：**
    >

>     > discardable-ratio在磁盘空间不足以同时保持titan和rocksdb数据时应该维持默认值。一般来说如果磁盘可用空间小于50%时推荐使用默认值。因为当discardable-ratio = 1.0时, Rocksdb数据一方面在不断增加，同时titan原有的blob文件回收需要该文件所有数据都迁移至RocksDB才会发生，这个过程会比较缓慢。discardable-ratio = 1.0带来的好处是减小compaction过程中Blob文件自身的GC，节省了带宽。

Reviewer: ```suggestion
    > `discardable-ratio` 在磁盘空间不足以同时保持 Titan 和 RocksDB 数据时应该维持默认值。一般来说，如果磁盘可用空间小于 50% 时，推荐使用默认值。因为当 `discardable-ratio = 1.0` 时，RocksDB 数据一方面在不断增加，同时 Titan 原有的 blob 文件回收需要该文件所有数据都迁移至 RocksDB 才会发生，这个过程会比较缓慢。`discardable-ratio = 1.0` 带来的好处是减小 compaction 过程中 Blob 文件自身的 GC，节省了带宽。
```

>     > `discardable-ratio` 在磁盘空间不足以同时保持 Titan 和 RocksDB 数据时应该维持默认值。一般来说，如果磁盘可用空间小于 50% 时，推荐使用默认值。因为当 `discardable-ratio = 1.0` 时，RocksDB 数据一方面在不断增加，同时 Titan 原有的 blob 文件回收需要该文件所有数据都迁移至 RocksDB 才会发生，这个过程会比较缓慢。`discardable-ratio = 1.0` 带来的好处是减小 compaction 过程中 Blob 文件自身的 GC，节省了带宽。

Reviewer: ```suggestion
    > `discardable-ratio` 在磁盘空间不足以同时保持 Titan 和 RocksDB 数据时应该维持默认值 `0.5`。一般来说，如果磁盘可用空间小于 50% 时，推荐使用默认值。因为当 `discardable-ratio = 1.0` 时，RocksDB 数据一方面在不断增加，同时 Titan 原有的 blob 文件回收需要该文件所有数据都迁移至 RocksDB 才会发生，这个过程会比较缓慢。`discardable-ratio = 1.0` 带来的好处是减小 compaction 过程中 Blob 文件自身的 GC，节省了带宽。
```

    enabled = false
    ```

### Titan 转 RocksDB 速度

> 由于 Titan Blob 文件中的 Value 是不连续的，而且 Titan 的 Cache 是 Value 级别，因此 Blob Cache 无法帮助 compaction。从 Titan 转到 RocksDB 速度相比 RocksDB 转 Titan 会慢一个数量级。在测试中，一个 800 GiB 的 TiKV 节点Titan 数据通过 tikv-ctl 做全量 compaction 转成 RocksDB，需要 12 个小时。

Reviewer: ```suggestion
由于 Titan Blob 文件中的 Value 是不连续的，而且 Titan 的 Cache 是 Value 级别，因此 Blob Cache 无法帮助 compaction。从 Titan 转到 RocksDB 速度相比 RocksDB 转 Titan 会慢一个数量级。在测试中，一个 800 GiB 的 TiKV 节点 Titan 数据通过 tikv-ctl 做全量 compaction 转成 RocksDB，需要 12 个小时。
```
